plugins/dell-dock/fu-dell-dock-usb-usb4.c

```diff
@@ -0,0 +1,658 @@
+					   buffer[self->blob_vid_major_offset],
+					   buffer[self->blob_vid_minor_offset]);
+	g_debug ("writing Thunderbolt vendor-id %s", dynamic_vendor_id);
+
+	/* get firmware blob product-id */
+	dynamic_product_id = g_strdup_printf ("%02x%02x",
+					   buffer[self->blob_pid_major_offset],
+					   buffer[self->blob_pid_minor_offset]);
+	g_debug ("writing Thunderbolt product-id %s", dynamic_product_id);
+
+	/* check vid, pid */
+	if (g_strcmp0 (self->nvm_vid, dynamic_vendor_id) != 0 ||
+	    g_strcmp0 (self->nvm_pid, dynamic_product_id) != 0) {
+		g_set_error (error, FWUPD_ERROR, FWUPD_ERROR_NOT_SUPPORTED,
+			     "Thunderbolt firmware vendor_id %s, product_id %s"
+			     "doesn't match NVM vendor_id %s, product_id %s",
+			     dynamic_vendor_id, dynamic_product_id,
+			     self->nvm_vid, self->nvm_pid);
+		return FALSE;
+	}
+
+	/* check header size */
+	memcpy (&header_offset, buffer, sizeof (guint32));
+	g_debug ("Header size: %" G_GUINT32_FORMAT, header_offset);
+	if (header_offset > image_size) {
+		g_set_error (error, FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE,
+			     "Image header is too big: %" G_GUINT32_FORMAT,
+			     header_offset);
+		return FALSE;
+	}
+
+	/* firmware install */
+	buffer += header_offset;
+	image_size -= header_offset;
+	if (!fu_dell_dock_usb4_hub_nvm_write(device, buffer, image_size, 0,
+					     error))
+		return FALSE;
+
+	fu_device_add_flag (device, FWUPD_DEVICE_FLAG_NEEDS_ACTIVATION);
+
+	/* dock will reboot to re-read; this is to appease the daemon */
+	//fu_device_set_status (device, FWUPD_STATUS_DEVICE_RESTART);
+	fu_device_set_version (device, dynamic_version);
+
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock_usb4_get_nvm_info (FuDevice *device, GError **error)
+{
+	FuDellDockUsb4 *self = FU_DELL_DOCK_USB4 (device);
+	guchar buf[NVM_READ_LENGTH] = {0};
+
+	if (!fu_dell_dock_usb4_hub_nvm_read (device, buf, NVM_READ_LENGTH, 0,
+					     error)) {
+		g_prefix_error (error, "NVM READ error:");
+		return FALSE;
+	}
+	self->nvm_ver = g_strdup_printf ("%02x.%02x",
+				         buf[NVM_VER_OFFSET_MAJOR],
+				         buf[NVM_VER_OFFSET_MINOR]);
+	self->nvm_vid = g_strdup_printf ("%02x%02x",
+				         buf[NVM_VID_OFFSET_MAJOR],
+				         buf[NVM_VID_OFFSET_MINOR]);
+	self->nvm_pid = g_strdup_printf ("%02x%02x",
+					 buf[NVM_PID_OFFSET_MAJOR],
+					 buf[NVM_PID_OFFSET_MINOR]);
+
+	fu_device_set_version_format (device, FWUPD_VERSION_FORMAT_PAIR);
+	fu_device_set_version (device, self->nvm_ver);
+	g_debug("NVM vendor_id: %s, product_id: %s", self->nvm_vid,
+						     self->nvm_pid);
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock_usb4_setup (FuDevice *device, GError **error)
+{
+	FuDellDockUsb4 *self = FU_DELL_DOCK_USB4 (device);
+	g_autofree const gchar *name;
+
+	if (!fu_dell_dock_usb4_get_nvm_info (device, error))
+		return FALSE;
+	name = g_strdup_printf ("TBT-%s%s", self->nvm_vid, self->nvm_pid);
+
+	fu_device_add_instance_id (device, name);
```

Comment:
If this code stays in dell-dock plugin, you need a check here that `self->nvm_vid` is `00d4` to make sure that non-Dell docks containing the same microcontroller for USB4 don't enumerate here.